File: code/ex329.java
class Solution {   
    public static int DFS(int[][] matrix, int i, int j, int l, int max, int m) {
        // Cannot be larger than max, no need to go further
        if (m - l <= max - l) {
            return max;
        }
        
        // Go left
        if (j > 0 && matrix[i][j - 1] > matrix[i][j])
            max = DFS(matrix, i, j - 1, l + 1, max, m);
        
        // Go right
        if (j < matrix[0].length - 1 && matrix[i][j + 1] > matrix[i][j])
            max = DFS(matrix, i, j + 1, l + 1, max, m);
    
        // Go up
        if (i > 0 && matrix[i - 1][j] > matrix[i][j])
            max = DFS(matrix, i - 1, j, l + 1, max, m);
        
        // Go down
        if (i < matrix.length - 1 && matrix[i + 1][j] > matrix[i][j])
            max = DFS(matrix, i + 1, j, l + 1, max, m);
        
        // Cannot go further
        max = Math.max(max, l);
        
            
        return max;
    }
    
    public int longestIncreasingPath(int[][] matrix) {
        int max = 0;
        int m = Integer.MIN_VALUE;
        
        for (int i = 0; i < matrix.length; i++)
            for (int j = 0; j < matrix[0].length; j++)
                m = Math.max(m, matrix[i][j]);

        
        for (int i = 0; i < matrix.length; i++)
            for (int j = 0; j < matrix[0].length; j++)
                max = DFS(matrix, i, j, 1, max, m);
    
        
        return max;
    }
}

File: code/ex329.cpp
class Solution {
    const int dir[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    int m, n;
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        m = matrix.size();    if (m == 0)    return 0;
        n = matrix[0].size();    if (n == 0)    return 0;
        vector<vector<int>> memo(m, vector<int>(n));
        int MAX = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                MAX = max(MAX, backtracking(matrix, memo, i, j));
            }
        }
        return MAX;
    }
    
private:
    int backtracking(vector<vector<int>>& matrix, vector<vector<int>>& memo, int i, int j) {
        if (memo[i][j] != 0)    return memo[i][j];
        
        int MAX = 1;
        for (auto& d : dir) {
            int ni = i + d[0], nj = j + d[1];
            if (ni < 0 || nj < 0 || ni >= m || nj >= n || matrix[ni][nj] <= matrix[i][j])
                continue;
            MAX = max(MAX, backtracking(matrix, memo, ni, nj) + 1);
        }
        
        memo[i][j] = MAX;
        return MAX;
    }
};